pkg/cpio/types.go

```diff
@@ -33,27 +34,61 @@ type RecordFormat interface {
 	Writer(w io.Writer) RecordWriter
 }
 
-func StaticRecord(contents string, info Info) Record {
+func StaticRecord(contents []byte, info Info) Record {
 	info.FileSize = uint64(len(contents))
 	return Record{
-		bytes.NewReader([]byte(contents)),
-		info,
+		ReadCloser: NewBytesReadCloser(contents),
+		Info:       info,
 	}
 }
 
-type eofReader struct{}
+type ReadCloser struct {
+	io.Reader
+}
 
-func (eofReader) Read(p []byte) (int, error) {
-	return 0, io.EOF
+func (b ReadCloser) Close() error {
+	return nil
+}
+
+func NewBytesReadCloser(contents []byte) io.ReadCloser {
+	return &ReadCloser{bytes.NewReader(contents)}
+}
+
+func NewReadCloser(r io.Reader) io.ReadCloser {
+	return &ReadCloser{r}
 }
 
 func EmptyRecord(info Info) Record {
 	return Record{
-		eofReader{},
+		NewBytesReadCloser([]byte{}),
 		info,
 	}
 }
 
+type DeferReader struct {
+	Name string
+	File *os.File
+}
+
+func (r *DeferReader) Read(p []byte) (int, error) {
+	if r.File == nil {
+		f, err := os.Open(r.Name)
+		if err != nil {
+			return -1, err
+		}
+		r.File = f
+	}
+	return r.File.Read(p)
+}
+
+func (r *DeferReader) Close() error {
+	return r.File.Close()
```

Comment:
I don't want to use it again - that's the point. If you set file = nil, another call to Read will just reopen the file instead of returning EOF again. Maybe this should set closed = true and Read should return EOF if the file is closed.